test/hack/bin/check-opsrecipes.sh

```diff
@@ -0,0 +1,143 @@
+#!/bin/bash
+set -euo pipefail
+
+# List of generated rules
+RULES_FILES=(./test/hack/output/*/prometheus-rules/templates/alerting-rules/*)
+#RULES_FILES=(./test/hack/output/*/prometheus-rules/templates/alerting-rules/up*)
+
+DEBUG_MODE=false
+CHECK_EXTRADATA_ERRORS=false
+CHECK_NORECIPE_ERRORS=true
+CHECK_UNEXISTINGRECIPE_ERRORS=true
+
+# Parameters:
+# - an element
+# - an array
+# Returns:
+# - 0 if element is not found in the array
+# - 1 if element is found in the array
+isInArray () {
+    referenceElement="$1" && shift
+    array=("$@")
+
+    for element in "${array[@]}"; do
+        [[ "$element" == "$referenceElement" ]] \
+            && return 0
+    done
+    return 1
+}
+
+listOpsRecipes () {
+    # find list of opsrecipes from git repo
+    tmpDir="$(mktemp -d)"
+    git clone --depth 1 --single-branch -b main -q git@github.com:giantswarm/giantswarm.git "$tmpDir"
+    # find all ops-recipes ".md" files, and keep only the opsrecipe name (may contain a path, like "rolling-nodes/rolling-nodes")
+    find "$tmpDir"/content/docs/support-and-ops/ops-recipes -type f -name \*.md \
+        | sed -n 's_'"$tmpDir"'/content/docs/support-and-ops/ops-recipes/\(.*\).md_\1_p'
+    rm -rf "$tmpDir"
+
+    # Add extra opsrecipes
+    # These ones are defined as aliases of `deployment-not-satisfied`:
+    echo "workload-cluster-managed-deployment-not-satisfied"
+    echo "workload-cluster-deployment-not-satisfied"
+    echo "deployment-not-satisfied-china"
+}
+
+main() {
+    local -a checkedRules
+    local -a opsRecipes
+    local -a E_extradata
+    local -a E_norecipe
+    local -a E_unexistingrecipe
+    local returncode=0
+
+    # Retrieve list of opsrecipes
+    mapfile -t opsRecipes < <(listOpsRecipes)
+
+    if [[ "$DEBUG_MODE" != "false" ]]; then
+        echo "List of opsrecipe:"
+        for recipe in "${opsRecipes[@]}"; do
+            echo " - \"$recipe\""
+        done
+    fi
+
+    # Look at each rules file
+    for rulesFile in "${RULES_FILES[@]}" ; do
+        # echo "rules file: $rulesFile"
+        prettyRulesFilename="$(basename "$rulesFile")"
+
+        # skip if rules file has already been checked
+        isInArray "$prettyRulesFilename" "${checkedRules[@]}" \
+            && continue
+
+        while read -r alertname opsrecipe severity overflow ; do
+
+            # Discard non-paging alerts
+            [[ "$severity" != "page" ]] && continue
+
+            # Get rid of anchors
+            opsrecipe="${opsrecipe%%#*}"
+            # Get rid of trailing slash
+            opsrecipe="${opsrecipe%/}"
+
+            if [[ "$overflow" != "" ]]; then
+                local message="file: $prettyRulesFilename / alert \"$alertname\" / recipe \"$opsrecipe\": extra data \"$overflow\""
+                E_extradata+=("$message")
+                continue
```

Comment:
What does this error mean ? What is even overflow ?

Reply:
I added comments, and even enabled the output of these errors.

Reply:
(and fixed a bug when there are no errors in one of these lists :grin: )